File: optimizer/src/chatbot/scenario/comparator.py
from __future__ import annotations

import pandas as pd


def _schedule_by_plan(schedule):
    if schedule is None or schedule.empty:
        return {}
    schedule = schedule.copy()
    schedule["plan_id"] = schedule["plan_id"].astype(str)
    return {str(row["plan_id"]): row for _, row in schedule.iterrows()}


def compare_scenarios(baseline_schedule, baseline_summary, scenario_schedule, scenario_summary):
    baseline = _schedule_by_plan(baseline_schedule)
    scenario = _schedule_by_plan(scenario_schedule)
    added = []
    removed = []
    rescheduled = []
    unchanged = []

    for plan_id in sorted(set(baseline) | set(scenario)):
        old = baseline.get(plan_id)
        new = scenario.get(plan_id)

        if old is None:
            added.append({"plan_id": plan_id, "field_id": str(new["field_id"]), "operation": str(new["operation"]), "start_time": str(new["start_time"])})
            continue

        if new is None:
            removed.append({"plan_id": plan_id, "field_id": str(old["field_id"]), "operation": str(old["operation"]), "start_time": str(old["start_time"])})
            continue

        old_candidate = str(old.get("candidate_id", ""))
        new_candidate = str(new.get("candidate_id", ""))
        old_machine = str(old.get("machine_id", ""))
        new_machine = str(new.get("machine_id", ""))

        if old_candidate != new_candidate or old_machine != new_machine:
            rescheduled.append({
                "plan_id": plan_id,
                "field_id": str(new["field_id"]),
                "operation": str(new["operation"]),
                "old_start_time": str(old["start_time"]),
                "new_start_time": str(new["start_time"]),
                "old_machine_id": old_machine,
                "new_machine_id": new_machine,
            })
        else:
            unchanged.append(plan_id)

    baseline_value = float(baseline_summary["total_objective_value_aud"])
    scenario_value = float(scenario_summary["total_objective_value_aud"])

    return {
        "baseline_objective_aud": round(baseline_value, 2),
        "scenario_objective_aud": round(scenario_value, 2),
        "objective_change_aud": round(scenario_value - baseline_value, 2),
        "baseline_direct_cash_effect_aud": float(baseline_summary["total_direct_cash_effect_aud"]),
        "scenario_direct_cash_effect_aud": float(scenario_summary["total_direct_cash_effect_aud"]),
        "baseline_terminal_value_aud": float(baseline_summary["total_terminal_value_aud"]),
        "scenario_terminal_value_aud": float(scenario_summary["total_terminal_value_aud"]),
        "actions_added": added,
        "actions_removed": removed,
        "actions_rescheduled": rescheduled,
        "unchanged_plan_count": len(unchanged),
    }

File: lib/optimizer/chatbot/scenario/comparator.ts
import { formatIso } from "../../datetime";
import type { OptimizationSummary, ScheduleRow } from "../../types";

interface ComparisonChange {
  plan_id: string;
  field_id: string;
  operation: string;
  start_time?: string;
  old_start_time?: string;
  new_start_time?: string;
  old_machine_id?: string;
  new_machine_id?: string;
}

export interface ScenarioComparison {
  baseline_objective_aud: number;
  scenario_objective_aud: number;
  objective_change_aud: number;
  baseline_direct_cash_effect_aud: number;
  scenario_direct_cash_effect_aud: number;
  baseline_terminal_value_aud: number;
  scenario_terminal_value_aud: number;
  actions_added: ComparisonChange[];
  actions_removed: ComparisonChange[];
  actions_rescheduled: ComparisonChange[];
  unchanged_plan_count: number;
}

function scheduleByPlan(schedule: ScheduleRow[] | null | undefined): Map<string, ScheduleRow> {
  const map = new Map<string, ScheduleRow>();
  if (!schedule) return map;
  for (const row of schedule) map.set(String(row.plan_id), row);
  return map;
}

export function compareScenarios(
  baselineSchedule: ScheduleRow[],
  baselineSummary: OptimizationSummary,
  scenarioSchedule: ScheduleRow[],
  scenarioSummary: OptimizationSummary,
): ScenarioComparison {
  const baseline = scheduleByPlan(baselineSchedule);
  const scenario = scheduleByPlan(scenarioSchedule);

  const added: ComparisonChange[] = [];
  const removed: ComparisonChange[] = [];
  const rescheduled: ComparisonChange[] = [];
  const unchanged: string[] = [];

  const allPlanIds = new Set([...baseline.keys(), ...scenario.keys()]);

  for (const planId of [...allPlanIds].sort()) {
    const oldRow = baseline.get(planId);
    const newRow = scenario.get(planId);

    if (!oldRow) {
      added.push({
        plan_id: planId,
        field_id: String(newRow!.field_id),
        operation: String(newRow!.operation),
        start_time: formatIso(newRow!.start_time),
      });
      continue;
    }

    if (!newRow) {
      removed.push({
        plan_id: planId,
        field_id: String(oldRow.field_id),
        operation: String(oldRow.operation),
        start_time: formatIso(oldRow.start_time),
      });
      continue;
    }

    const oldCandidate = String(oldRow.candidate_id ?? "");
    const newCandidate = String(newRow.candidate_id ?? "");
    const oldMachine = String(oldRow.machine_id ?? "");
    const newMachine = String(newRow.machine_id ?? "");

    if (oldCandidate !== newCandidate || oldMachine !== newMachine) {
      rescheduled.push({
        plan_id: planId,
        field_id: String(newRow.field_id),
        operation: String(newRow.operation),
        old_start_time: formatIso(oldRow.start_time),
        new_start_time: formatIso(newRow.start_time),
        old_machine_id: oldMachine,
        new_machine_id: newMachine,
      });
    } else {
      unchanged.push(planId);
    }
  }

  const baselineValue = baselineSummary.total_objective_value_aud;
  const scenarioValue = scenarioSummary.total_objective_value_aud;

  return {
    baseline_objective_aud: Math.round(baselineValue * 100) / 100,
    scenario_objective_aud: Math.round(scenarioValue * 100) / 100,
    objective_change_aud: Math.round((scenarioValue - baselineValue) * 100) / 100,
    baseline_direct_cash_effect_aud: baselineSummary.total_direct_cash_effect_aud,
    scenario_direct_cash_effect_aud: scenarioSummary.total_direct_cash_effect_aud,
    baseline_terminal_value_aud: baselineSummary.total_terminal_value_aud,
    scenario_terminal_value_aud: scenarioSummary.total_terminal_value_aud,
    actions_added: added,
    actions_removed: removed,
    actions_rescheduled: rescheduled,
    unchanged_plan_count: unchanged.length,
  };
}
